packages/calypso-e2e/src/lib/pages/site-editor-page.ts

```diff
@@ -0,0 +1,170 @@
+ */
+export class SiteEditorPage {
+	private page: Page;
+
+	private editorToolbarComponent: EditorToolbarComponent;
+	private editorGutenbergComponent: EditorGutenbergComponent;
+	private editorSidebarBlockInserterComponent: EditorSidebarBlockInserterComponent;
+	private editorWelcomeTourComponent: EditorWelcomeTourComponent;
+
+	/**
+	 * Constructs an instance of the page POM class.
+	 *
+	 * @param {Page} page The underlying page.
+	 */
+	constructor( page: Page ) {
+		this.page = page;
+
+		this.editorToolbarComponent = new EditorToolbarComponent( page, this.editor );
+		this.editorWelcomeTourComponent = new EditorWelcomeTourComponent( page, this.editor );
+		this.editorSidebarBlockInserterComponent = new EditorSidebarBlockInserterComponent(
+			page,
+			this.editor
+		);
+		// Because of the unique extra iframe in the site editor, this component needs the canvas locator.
+		this.editorGutenbergComponent = new EditorGutenbergComponent( page, this.editorCanvas );
+	}
+
+	/**
+	 * A frame-safe locator to the top level element in the editor, for building other editor locators.
+	 */
+	private get editor(): Locator {
+		return this.page.url().includes( wpAdminPath )
+			? this.page.locator( selectors.editorRoot )
+			: this.page.frameLocator( selectors.editorIframe ).locator( selectors.editorRoot );
+	}
+
+	/**
+	 * A frame-safe locator to the top level element in the editor canvas, for building other editor locators.
+	 * The editor canvas is an iframe wrapping the center editor area that is unique to the site-editor.
+	 */
+	private get editorCanvas(): Locator {
+		return this.editor
+			.frameLocator( selectors.editorCanvasIframe )
+			.locator( selectors.editorCanvasRoot );
+	}
+
+	/**
+	 * Visit the site editor by URL directly.
+	 *
+	 * @param siteHostName Host name of the site, without scheme. (e.g. testsite.wordpress.com)
+	 */
+	async visit( siteHostName: string ): Promise< void > {
+		await this.page.goto( getCalypsoURL( `site-editor/${ siteHostName }` ) );
+	}
+
+	/**
+	 * Waits until the site editor is fully loaded.
+	 */
+	async waitUntilLoaded(): Promise< void > {
+		// There are more stages to the site editor loading than the regular editor.
+		// The most reliable "last" thing to load is the canvas iframe
+		await this.editorCanvas.waitFor( { timeout: 60 * 1000 } );
+		// But then, template parts load async afterwards!
+		const spinnerLocator = this.editorCanvas.locator( selectors.templateLoadingSpinner );
+		// There could be many spinners, so we will keep waiting for the first to be detached.
+		await spinnerLocator.first().waitFor( { state: 'detached' } );
+	}
+
+	/**
+	 * Does all waiting and initial actions to prepare the site editor for interaction.
+	 *
+	 * @param {object} param0 Keyed object of options.
+	 * @param {boolean} param0.leaveWithoutSaving Set if we should auto-except dialog about unsaved changes when leaving.
+	 */
+	async prepareForInteraction(
```

Comment:
I'm a bit confused as to the reason for this to method to exist when the contents of this can be incorporated into `waitUntilLoaded`.

The way I understand it, this method adds dialog handling too, but I feel like it should be incorporated into the individual actions that are expecting to cause the popup. In this case, I would have added dialog handling handling to the `redo` and `undo` methods instead.

Reply:
What I am imagining is something like this:

```
async redo( { dismissPopup: boolean = false } )
    // ...popup handling
```

Reply:
Ahh sure, those are great points!

>I'm a bit confused as to the reason for this to method to exist when the contents of this can be incorporated into waitUntilLoaded.

So, the main intent here was separation of concerns, and not hiding assumptions in `waitUntilLoaded`.

`prepareForInteraction` does two additional things after `waitUntilLoaded`:
1. Dismisses the welcome tour.
2. Optionally sets up a handler that lets you exit the whole editor without saving your changes.

The way I see it, those two actions are _common_ things script writers may want to do, but they are not things we may want to do in _every_ case. For example, a script in the future may test the welcome tour, or we may have a saving-based script that wouldn't want to auto-dismiss the saving warning dialog.

To me, putting those in `waitUntilLoaded` felt misleading, because they are technically more than just waiting for the editor to finish loading.

>I feel like it should be incorporated into the individual actions that are expecting to cause the popup. In this case, I would have added dialog handling to the redo and undo methods instead.

I very much agree with this design approach! But in this case, the dialog we are targeting is the "leave without saving" one that comes from leaving the editor. Because this isn't related to any specific action (like redo/undo), but rather the editor state as a whole, it felt like it should be defined in a place that addresses the overall configuration of the editor.

Does that make sense?

Reply:
> Because this isn't related to any specific action (like redo/undo), but rather the editor state as a whole, it felt like it should be defined in a place that addresses the overall configuration of the editor.

Right, I had thought the action of redo/undo was what triggers the popup. If not, then the placement for the interaction makes sense.

We will iterate as needed on the name or placement of these calls as we work on this framework.